File: solutions/haskell/luhn/1/src/Luhn.hs
module Luhn (isValid) where

isValid :: String -> Bool
isValid strNum = (validate . toInt . filt) strNum
    where
        filt :: String -> String
        filt = filter (\ chr -> chr >= '0' && chr <= '9')
        
        toInt :: String -> [Integer]
        toInt = map (\chr -> (read [chr]) :: Integer)

        validate :: [Integer] -> Bool
        validate [] = False
        validate [_] = False
        validate nums = mod (suma (if (even $ length nums) then nums else 0:nums)) 10 == 0
            where
                suma :: [Integer] -> Integer 
                suma nums = sum $ zipWith (\num fun -> fun num) nums $ cycle [(\a -> if a<=4 then 2*a else 2*a-9), id]

File: solutions/julia/luhn/1/luhn.jl
function luhn(str0::String) :: Bool
    nospaces = filter((ch->ch!=' '),str0)
    if any(ch->!isnumeric(ch),nospaces) return false end
    numbers = [chr-'0' for chr in nospaces]
    if length(numbers)<=1 return false end
    if isodd(length(numbers)) numbers=[0;numbers] end
    for i in 1:2:(length(numbers))
        n=numbers[i]*2
        numbers[i]=n>9 ? n-9 : n
    end
    @show numbers
    return 0==mod(sum(numbers),10)
end